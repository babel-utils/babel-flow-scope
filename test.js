// @flow
'use strict';

const pluginTester = require('babel-plugin-tester');
const getBabylonOptions = require('babylon-options');
const printAST = require('ast-pretty-print');
const {getFlowBindingsInScope} = require('./');

const babelOptions = {
  parserOpts: getBabylonOptions({
    stage: 0,
    plugins: ['flow'],
  }),
};

const plugin = ({ types: t }) => {
  let getOutput = path => {
    let bindings = getFlowBindingsInScope(path);
    let printed = printAST(bindings) + '\n';
    return t.expressionStatement(
      t.templateLiteral([
        t.templateElement({ raw: printed }, true)
      ], [])
    );
  };

  return {
    name: 'test-plugin',
    visitor: {
      Program(path) {
        path.unshiftContainer('body', getOutput(path));
      },
      'ClassDeclaration|FunctionDeclaration|InterfaceDeclaration|TypeAlias'(path) {
        path.insertAfter(getOutput(path));
      },
      ClassExpression(path) {
        path.find(p => p.isDeclaration()).insertAfter(getOutput(path));
      }
    },
  };
};

pluginTester({
  plugin,
  snapshot: true,
  babelOptions,
  tests: {
    'import type': 'import type a from "mod";',
    'import {type}': 'import {type a} from "mod";',
    'import typeof': 'import typeof {a} from "mod";',
    'import {typeof}': 'import {typeof a} from "mod";',
    'type alias': 'type a = {};',
    'interface declaration': 'interface a {}',
    'class declaration': 'class a {}',
    'type alias params': 'type a<b> = {}',
    'interface params': 'interface a<b> {}',
    'class params': 'class a<b> {}',
    'function params': 'function a<b>() {}',
    'nested': 'function a<b>() {}',
    'class expression': 'let a = class b {}',
  }
})
