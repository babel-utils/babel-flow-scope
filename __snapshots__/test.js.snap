// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1. import type 1`] = `
"
import type a from \\"mod\\";

      ↓ ↓ ↓ ↓ ↓ ↓

\`
  a:
    kind: 'import'
    path: Path: Identifier (1:12,1:13)
      name: 'a'
\`;
import type a from \\"mod\\";
"
`;

exports[`2. import {type} 1`] = `
"
import {type a} from \\"mod\\";

      ↓ ↓ ↓ ↓ ↓ ↓

\`
  a:
    kind: 'import'
    path: Path: Identifier (1:13,1:14)
      __clone: Function
      name: 'a'
\`;
import { type a } from \\"mod\\";
"
`;

exports[`3. import typeof 1`] = `
"
import typeof {a} from \\"mod\\";

      ↓ ↓ ↓ ↓ ↓ ↓

\`
  a:
    kind: 'import'
    path: Path: Identifier (1:15,1:16)
      __clone: Function
      name: 'a'
\`;
import typeof { a } from \\"mod\\";
"
`;

exports[`4. import {typeof} 1`] = `
"
import {typeof a} from \\"mod\\";

      ↓ ↓ ↓ ↓ ↓ ↓

\`
  a:
    kind: 'import'
    path: Path: Identifier (1:15,1:16)
      __clone: Function
      name: 'a'
\`;
import { typeof a } from \\"mod\\";
"
`;

exports[`5. type alias 1`] = `
"
type a = {};

      ↓ ↓ ↓ ↓ ↓ ↓

\`
  a:
    kind: 'declaration'
    path: Path: Identifier (1:5,1:6)
      name: 'a'
\`;
type a = {};
\`{}
\`;
"
`;

exports[`6. interface declaration 1`] = `
"
interface a {}

      ↓ ↓ ↓ ↓ ↓ ↓

\`
  a:
    kind: 'declaration'
    path: Path: Identifier (1:10,1:11)
      name: 'a'
\`;
interface a {}
\`{}
\`;
"
`;

exports[`7. type alias params 1`] = `
"
type a<b> = {}

      ↓ ↓ ↓ ↓ ↓ ↓

\`
  a:
    kind: 'declaration'
    path: Path: Identifier (1:5,1:6)
      name: 'a'
\`;
type a<b> = {};
\`
  b:
    kind: 'param'
    path: Path: TypeParameter (1:7,1:8)
      bound: undefined
      name: 'b'
      variance: null
\`;
"
`;

exports[`8. interface params 1`] = `
"
interface a<b> {}

      ↓ ↓ ↓ ↓ ↓ ↓

\`
  a:
    kind: 'declaration'
    path: Path: Identifier (1:10,1:11)
      name: 'a'
\`;
interface a<b> {}
\`
  b:
    kind: 'param'
    path: Path: TypeParameter (1:12,1:13)
      bound: undefined
      name: 'b'
      variance: null
\`;
"
`;

exports[`9. class params 1`] = `
"
class a<b> {}

      ↓ ↓ ↓ ↓ ↓ ↓

\`{}
\`;
class a<b> {}
\`
  b:
    kind: 'param'
    path: Path: TypeParameter (1:8,1:9)
      bound: undefined
      name: 'b'
      variance: null
\`;
"
`;

exports[`10. function params 1`] = `
"
function a<b>() {}

      ↓ ↓ ↓ ↓ ↓ ↓

\`{}
\`;
function a<b>() {}
\`
  b:
    kind: 'param'
    path: Path: TypeParameter (1:11,1:12)
      bound: undefined
      name: 'b'
      variance: null
\`;
"
`;

exports[`11. nested 1`] = `
"
function a<b>() {}

      ↓ ↓ ↓ ↓ ↓ ↓

\`{}
\`;
function a<b>() {}
\`
  b:
    kind: 'param'
    path: Path: TypeParameter (1:11,1:12)
      bound: undefined
      name: 'b'
      variance: null
\`;
"
`;
